### PR TITLE
Add summoner

### PR DESCRIPTION
**Pull request by jung1411**
# Before putting PR (put x in bracket if you ran it.);

- [x] Run `pylint $(git ls-files '*.py')` passes

# Overview
https://github.com/dhmoon91/discord/issues/4
Add new summoners to the local json file and send the list of summoners and their rank information to discord bot.
- Input can be written with (comma, space or comma + space) between each summoner names
- Creates data.json file if it doesn't exist and updates the file if it exists with new summoner
- Checks whether the summoner name exists in API and sends discord bot error if it does not exist

Excluded:
- I think delete data should be a separate function/issue and I have made an issue regarding remove/clear command (https://github.com/dhmoon91/discord)

When adding summoners:
Desktop View:
image

Mobile View:
image

When total number of summoners exceed 10 people:
Desktop View:
image

Mobile View:
image

When adding summoner that does not exist:
Desktop View:
image

Mobile View:
image

Listing the summoners added in the server:
image

### CODE REVIEW

**Comment by dhmoon91** [comment]
`data.json` should not be commited. Each environments will have different data set

File: bot.py [hunk 1]
```diff
@@ -99,6 +102,97 @@ async def get_last_match(ctx, name: str):
...
+    if os.path.getsize(json_path) > 0:
+        with open(json_path, "r") as file:
+            file_data = json.load(file)
+
+        # if server id exist in json, add number of players
+        if server_id in file_data:
+            number_of_players += len(file_data[server_id])
+
+    # make dictionary for newly coming in players
+    players_list_info = create_summoner_list(players_list, server_id)
```
Comment (dhmoon91) [inline bot.py]: You should only call API for 'data not pulled' summoners. 
This is going to call API even after we already pulled info in previous command

Comment (jung1411) [reply bot.py]: fixed only call API for not pulled summoners with new push.
I don't know why pylint gives 10/10 when I run it locally but gives error when pylint checks on pr. So I just pushed again after fixing them.

File: bot.py [hunk 2]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
     os.remove("df_styled.png")
 
 
+@bot.command(name="add", help="Add the players to the list")
+async def add_summoner(ctx, *, message):
+    """Writes list of summoners to local
+    json file and sends the list to the bot"""
+
+    async with ctx.typing():
+        await asyncio.sleep(1)
```
Comment (dhmoon91) [inline bot.py]: what does this line do?

File: bot.py [hunk 3]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
...
+
+    json_path = "data/data.json"
```
Comment (dhmoon91) [inline bot.py]: this is gonna break on prod.

File: riot.py [hunk 4]
```diff
@@ -109,3 +110,87 @@ def previous_match(name: str):
...
+
+        # check if there is any rank data on user as unranked players don't have any data on ranked
+        if len(ranked_stat) > 0:
+
+            solo_rank_stat = pydash.find(ranked_stat, {"queueType": "RANKED_SOLO_5x5"})
+
+            tier_division = solo_rank_stat["tier"]
+            tier_rank = solo_rank_stat["rank"]
+
+            tier_rank_map = {"I": "1", "II": "2", "III": "3", "IV": "4"}
```
Comment (dhmoon91) [inline riot.py]: this should be ported out. We will be using this in different functions too.

File: riot.py [hunk 5]
```diff
@@ -109,3 +110,87 @@ def previous_match(name: str):
...
+    Returns:
+    boolean (bool): if doesn't exist, boolean of whether if the summoner exists in API
+    """
+
+    try:
+        watcher.summoner.by_name(MY_REGION, summoner)
+    except requests.exceptions.HTTPError:
+        return False
+
+    return True
```
Comment (dhmoon91) [inline riot.py]: you are not checking anything here. 

What are data needed for us to say this summoner exists in NA LOL data? If league provides us with summoner name, is that good enough for us to make a team?

You should check thoese params after watcher.summoner.by_name(MY_REGION, summoner) then return true there. Right now you are just depending on league API to throw error for you.

File: bot.py [hunk 6]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
...
+        with open(json_path, "r") as file:
+            file_data = json.load(file)
+
+        # if server id exist in json, add number of players
+        if server_id in file_data:
+            number_of_players += len(file_data[server_id])
+
+    # send error message to bot then exits out of the function if an error with summoner's name
+    for count, _ in enumerate(players_list):
+        if not check_summoner_name(players_list[count]):
```
Comment (dhmoon91) [inline bot.py]: Again, this is calling API even if user command includes already added summoner name.
You should first do;
1) Check if given summoner name exists in `data` fiel
2) Check if we have enough info like tier and rank info in our data

Comment (dhmoon91) [reply bot.py]: Also, what if we already have 10 players in our data? before this call?

File: bot.py [hunk 7]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
...
+            players_list.remove(players_list[count])
+
+    # add number of players from incoming data
+    number_of_players += len(players_list)
+
+    # if more than 10 players, send error message
+    if number_of_players > 10:
+        num_of_players_needed = 10 - len(file_data[server_id])
+        await ctx.send("You have exceeded limit of 10 summoners!\n")
+        await ctx.send(f"Please add {num_of_players_needed} more summoners!\n")
```
Comment (dhmoon91) [inline bot.py]: If case matches and this should end the call, just return here. Then you don't need another `else` on next line.

File: bot.py [hunk 8]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
...
+                name="** **",
+                value=f"Adding multiple summoners:\n`@{bot.user.name} add name1, name2`",
+                inline=False,
+            )
+
+            await ctx.send(embed=embed)
+            return None
+
+        # changes the incoming summoner names to what's in the api
+        players_list[count] = get_summoner_name(players_list[count])
```
Comment (dhmoon91) [inline bot.py]: why is this needed?

**Comment by dhmoon91** [comment]
I don't get why you have multiple functions that's basically doing same job over and over `check_summoner_name` `get_summoner_name` `create_summoner_list`.

You are basically requesting same data like `summoner name` 3 times over this one `add` command. THIS IS BAD DESIGN - What if later we are required to pay for each API call?

Please think this through again and redesign it out. You should have one function that gets all necessary data in one flow, then throw from that function if certain criteria doesn't look right.

File: bot.py [hunk 9]
```diff
@@ -99,6 +110,131 @@ async def get_last_match(ctx, name: str):
...
+            players_list[count] in player["user_name"]
+            for player in file_data[server_id]
+        ):
+            players_list.remove(players_list[count])
+
+    # add number of players from incoming data
+    number_of_players += len(players_list)
+
+    # if more than 10 players, send error message
+    if number_of_players > 10:
```
Comment (dhmoon91) [inline bot.py]: The `10` should be used as constant; `MAX_NUM_PLAYERS_TEAM=10`.

File: bot.py (hunk 2 above)
Comment (jung1411) [reply bot.py]: It's for typing indicator like when you run !add command, at the bottom it shows @관전남 is typing...
I'll put a comment there.

File: bot.py (hunk 7 above)
Comment (jung1411) [reply bot.py]: So you don't want it to show the list of summoners added so far if it exceeds 10 people?
Just the meesage?

File: bot.py (hunk 8 above)
Comment (jung1411) [reply bot.py]: Say if we input a summoner name "lifeissohard", but then the actual league name will be "Life is so hard" and if we want to check if the summoner name exists in our json file, the name has to match.

File: bot.py (hunk 7 above)
Comment (dhmoon91) [reply bot.py]: sync'd on call - Print out `current list of summoners` should be ported out as separate function and called upon error + success

**Comment by HelloFive** [comment]
Screen Shot 2021-07-04 at 5 34 28 AM
Clean up git commits before merging PR. Ideally merge all of commits into 1-2 commits with more meaningful commits for this PR.

**Comment by dhmoon91** [comment]
> Screen Shot 2021-07-04 at 5 34 28 AM
> 
> Clean up git commits before merging PR. Ideally merge all of commits into 1-2 commits with more meaningful commits for this PR.

yeah this or we can just squash the commits upon merging.

File: bot.py (hunk 3 above)
Comment (HelloFive) [reply bot.py]: I guess absolute path is needed here?

File: requirements.txt [hunk 10]
```diff
@@ -6,4 +6,6 @@ pandas
 riotwatcher
 dataframe_image
 black
-pylint
\ No newline at end of file
+pylint
+requests
```
Comment (dhmoon91) [inline requirements.txt]: Where does `requests` package gets used?

File: bot.py [hunk 11]
```diff
@@ -162,6 +168,124 @@ async def get_last_match(ctx, name: str):
...
+
+        embed_data.fields = []
+        embed_data.fields.append(
+            {"name": "Summoners", "value": output_str, "inline": False}
+        )
+
+        await ctx.send(embed=create_embed(embed_data))
+    # pylint: disable=broad-except
+    except Exception as e_values:
+        print(str(e_values))
```
Comment (dhmoon91) [inline bot.py]: remove `print`

File: bot.py [hunk 12]
```diff
@@ -162,6 +168,124 @@ async def get_last_match(ctx, name: str):
...
+
+        if "404" in str(e_values):
+            error_title = "Invalid Summoner Name"
+            error_description = "You have entered an invalid summoner name!"
+        elif not str(e_values):
```
Comment (dhmoon91) [inline bot.py]: you are just checking if `e_values` not string here? what if other error comes out that is not related to `limit of 10 summoners`?

**Review by dhmoon91 — APPROVED** [review]
@HelloFive I'm approving this to unblock you so this can be merged and you can start working on it. 
But give this a look as well I think you should take over this PR as well and refactor or make improvements -> merge -> move on to making teams

File: riot.py [hunk 13]
```diff
@@ -115,3 +114,60 @@ def previous_match(name: str):
...
+                    "formatted_user_name": user_name,
+                    "tier_division": tier_division,
+                    "tier_rank_number": tier_rank_number,
+                }
+            )
+
+        return members_to_add
+    # pylint: disable=broad-except
+    except Exception as e_values:
+        print(e_values.args)
```
Comment (dhmoon91) [inline riot.py]: remove print

Comment (jjaykim) [reply riot.py]: Do you `need try and except` here?
I think the `except` in bot.py will catch the error with message when summoner name not found from API

**Comment by HelloFive** [comment]
@jung1411 commit messages with "WIP" should be cleaned.
I suggest fixup or reword it to something meaningful.
Screen Shot 2021-07-06 at 10 57 51 PM

File: requirements.txt (hunk 10 above)
Comment (jung1411) [reply requirements.txt]: done

File: bot.py (hunk 11 above)
Comment (jung1411) [reply bot.py]: done

File: bot.py (hunk 12 above)
Comment (jung1411) [reply bot.py]: created an error name and caught error type in exception
e.g.  `elif "Limit Exceeded" in str(e_values):`

File: riot.py (hunk 13 above)
Comment (jung1411) [reply riot.py]: done

File: bot.py (hunk 3 above)
Comment (HelloFive) [reply bot.py]: Resolved with
```
data_folder_path = get_file_path("data/")
json_path = data_folder_path + "data.json"
```

File: riot.py (hunk 4 above)
Comment (HelloFive) [reply riot.py]: Resolved by defining in separate file: `utils/constants.py`

File: bot.py [hunk 14]
```diff
@@ -162,6 +168,123 @@ async def get_last_match(ctx, name: str):
...
+        # typing indicator
+        async with ctx.typing():
+            await asyncio.sleep(1)
+
+        data_folder_path = get_file_path("data/")
+        json_path = data_folder_path + "data.json"
+
+        # create a directory containing json file to store data for added summoners
+        if not os.path.exists(json_path):
+            os.makedirs(data_folder_path)
```
Comment (HelloFive) [inline bot.py]: also check for when `data_folder_path` exists or not
If the folder exists, we don't need to create folder.
I was hitting issue in this case.

Comment (HelloFive) [reply bot.py]: Try using this `os.makedirs(path, exist_ok = True)`
Reference: https://www.geeksforgeeks.org/python-os-makedirs-method/

Comment (HelloFive) [reply bot.py]: Steps to reproduce Issue I hit:
1. delete data.json folder
2. run add command with 관전남
3. I got error such as https://discord.com/channels/853739164491317298/859326156859899924/863609285216305194

Solution was to `rm -rf data` folder itself, but can be solved in code with above change.

File: bot.py [hunk 15]
```diff
@@ -162,6 +168,123 @@ async def get_last_match(ctx, name: str):
...
+                    for player in file_data[server_id]
+                ):
+                    player_list.remove(player_name)
+
+        # add number of summoners from incoming data to total number of players
+        total_number_of_players += len(player_list)
+
+        if total_number_of_players > MAX_NUM_PLAYERS_TEAM:
+            raise Exception(
+                "Limit Exceeded", "You have exceeded a limit of 10 summoners!"
```
Comment (HelloFive) [inline bot.py]: I might be being little picky, but I liked previous way of returning how many more players needs to be added.
`num_of_players_needed = 10 - len(file_data[server_id])`

File: bot.py [hunk 16]
```diff
@@ -162,6 +168,123 @@ async def get_last_match(ctx, name: str):
...
+        if "404" in str(e_values):
+            error_title = "Invalid Summoner Name"
+            error_description = f"`{e_values.args[1]}` is not a valid name. \
+                \n\nAdding multiple summoners:\n `@{bot.user.name} add name1, name2`"
+        elif "Limit Exceeded" in str(e_values):
+            error_title = "Unable to Add"
+            error_description = f"{e_values.args[1]}"
+        else:
+            error_title = "Error"
+            error_description = "Oops! Something went wrong.\nTry again!"
```
Comment (HelloFive) [inline bot.py]: Could we output `e_values` here as well?
We currently do not have logs available for debug, having e_values here can help to troubleshoot any other issue other than `404` or `Limit Exceeded`

File: bot.py (hunk 7 above)
Comment (HelloFive) [reply bot.py]: @jung1411 looks like issue for 1st comment has been resolved by 
```
if total_number_of_players > MAX_NUM_PLAYERS_TEAM:
            raise Exception(
                "Limit Exceeded", "You have exceeded a limit of 10 summoners!"
            )
```
But for the 3rd comment, it's not resolved yet.
When I exceed more than 10 players, I only get error message without display of `current list of summoners`
Screen Shot 2021-07-10 at 11 17 21 PM

File: bot.py (hunk 16 above)
Comment (jung1411) [reply bot.py]: 아아 동환이가 404랑 10명넘는거 외로는 그냥 general로 받으라고하긴했었는데
그리고 일단 embed로 보낼 info여서 specific info가 필요없긴할듯

Comment (HelloFive) [reply bot.py]: 내가 생각한건, 
```
error_description = f"Oops! Something went wrong.Try again!\n{e_values.args[1]}"
```
이거였는데, so that we can see what unexpected error message is.
근데 이렇게 할시에 문제점은 customer is not suppose to see such error message. 그래서 굳이 안 고쳐도될듯.

File: bot.py [hunk 17]
```diff
@@ -209,45 +209,110 @@ async def add_summoner(ctx, *, message):
...
+async def display_current_list_of_summoners(ctx):
+    """For displaying current list of summoners"""
+
+    try:
+        # server id
+        server_id = str(ctx.guild.id)
+
+        # folder and path for data json
+        data_folder_path = get_file_path("data/")
+        json_path = data_folder_path + "data.json"
```
Comment (HelloFive) [inline bot.py]: looks like `data_folder_path` and `json_path` has been repeatedly.
Please set the variables once and re-use it.

**Review by HelloFive — APPROVED** [review]
LGTM
